draft-ietf-taps-interface.md

```diff
@@ -1784,37 +1775,183 @@ that the other endpoint is done sending on a connection.
-Deframe. Since the Deframer depends on the protocol used at the application
-layer, it is bound to the Preconnection during the pre-establishment phase:
+A Message Framer is primarily defined by the set of code that handles events
+for the Framer, specifically how it handles inbound and outbound data
+parsing.
 
+Applications can instantiate a Message Framer upon which they will receive
+framing events, or use a Message Framer defined by another library.
+
+~~~
+framer := NewMessageFramer()
 ~~~
-Preconnection.DeframeWith(Deframer)
 
-{messageData} := Deframer.Deframe(OctetStream)
+## Adding Message Framers to Connections
+
+The Message Framer object can then be added to one or more Preconnections
+to run on top of transport protocols. Multiple Framers may be added. If multiple
+Framers are added, the last one added runs first when framing outbound messages,
+and last when parsing inbound data.
+
 ~~~
+Preconnection.AddFramer(framer)
+~~~
+
+When sending Messages, applications can add specific Message
+values to a MessageContext ({{message-props}}) that is intended for a Framer.
+This can be used, for example, to set the type of a Message for a TLV format.
+The namespace of values is custom for each unique Message Framer.
+
+~~~
+messageContext := NewMessageContext()
+messageContext.add(framer, key, value)
+Connection.Send(messageData, messageContext)
+~~~
+
+When an application receives a MessageContext in a Receive event,
+it can also look to see if a value was set by a specific Message Framer.
+
+~~~
+messageContext.get(framer, key) -> value
+~~~
+
+## Message Framer Lifetime
+
+When a Connection establishment attempt begins, an event is delivered to
+notify the Framer that a new Connection is being created.
+Similarly, a stop event is delivered when a Connection is being torn down.
+The Framer can use the Connection object to look up specific properties
+of the connection or the network being used that may influence how
+to frame Messages.
+
+~~~
+MessageFramer -> Start(Connection)
+MessageFramer -> Stop(Connection)
+~~~
+
+When Message Framer receives a `Start` event, it has the opportunity to start
+writing some data on the Connection prior to the Connection delivering its
+`Ready` event. This allows the Framer to communicate control data to the
+remote endpoint that can be used to parse Messages.
+
+~~~
+MessageFramer.MakeConnectionReady(Connection)
+~~~
+
+At any time if the Message Framer encounters a fatal error, it can also cause the Connection
+to fail and provide an error.
+
+~~~
+MessageFramer.FailConnection(Connection, Error)
+~~~
+
+Before a Message Framer marks itself as ready, it can also dynamically add a protocol or framer
+above it in the stack. This allows protocols like STARTTLS, that need to add TLS conditionally,
+to modify the Protocol Stack based on a handshake result.
+
+~~~
+otherFramer := NewMessageFramer()
+MessageFramer.PrependFramer(Connection, otherFramer)
+~~~
```

Comment:
I think, since handshakes build from the bottom up, and you don't want outstanding data in the stream below you, it's only safe to prepend.

Reply:
Right. So if a would do some kind of multiplexing within my framer, i can not easily reuse other framers that further process my messages. Anyway, this would require to dispatch stuff to another framer, not appending it.